pkg/cmd/pulumi/templates/org.go

```diff
@@ -0,0 +1,221 @@
+)
+
+func (s *Source) getOrgTemplates(
+	ctx context.Context, templateName string,
+	interactive bool, wg *sync.WaitGroup,
+) {
+	ws := pkgWorkspace.Instance
+	project, _, err := ws.ReadProject()
+	if err != nil && !errors.Is(err, workspace.ErrProjectNotFound) {
+		s.addError(err)
+		return
+	}
+
+	b, err := cmdBackend.CurrentBackend(ctx, ws, cmdBackend.DefaultLoginManager, project, display.Options{
+		Color:         cmdutil.GetGlobalColorization(),
+		IsInteractive: interactive,
+	})
+	if err != nil {
+		if !errors.Is(err, backend.MissingEnvVarForNonInteractiveError{}) {
+			s.addError(err)
+		}
+		return
+	}
+	if !b.SupportsTemplates() {
+		return
+	}
+
+	logging.Infof("Listing Org Templates from the cloud")
+	user, orgs, _, err := b.CurrentUser()
+	if err != nil {
+		s.addError(err)
+		return
+	} else if user == "" {
+		return // No current user - so don't proceed.
+	}
+
+	alreadySeenSourceURLs := map[string]struct{}{}
+
+	handleOrg := func(org string) {
+		defer wg.Done()
+		logging.Infof("Checking for templates from %q", org)
+		orgTemplates, err := b.ListTemplates(ctx, org)
+		if apiError := new(apitype.ErrorResponse); errors.As(err, &apiError) {
+			// This is what happens when we try to access org templates for an org that hasn't enabled org templates.
+			if apiError.Code == 402 {
+				logging.Infof("%q does not have access to org templates (code=%d)", org, apiError.Code)
+				return
+			}
+		} else if err != nil {
+			s.addError(err)
+			logging.Warningf("Failed to get templates from %q: %s", org, err.Error())
+			return
+		} else if orgTemplates.HasAccessError {
+			logging.Warningf("Failed to get templates from %q: Access Denied", org)
+			return
+		} else if orgTemplates.HasUpstreamError {
+			logging.Warningf("Failed to get templates from %q: Upstream Error", org)
+			return
+		}
+
+		for source, sourceTemplates := range orgTemplates.Templates {
+			logging.Infof("sourcing templates from %q", source)
+			for _, template := range sourceTemplates {
+				// These template are maintained using https://github.com/pulumi/templates, and are
+				// ingested without going through the Pulumi Cloud.
+				//
+				//
+				if strings.HasPrefix(template.TemplateURL, "https://github.com/pulumi/templates") {
+					continue
+				}
+
+				// Check if we already have this template from another source.
+				if _, ok := alreadySeenSourceURLs[template.TemplateURL]; ok {
+					// Skip a template that we have already seen.
+					continue
+				}
+				alreadySeenSourceURLs[template.TemplateURL] = struct{}{}
+
+				// If we are searching for a template of a specific name,
+				// only match templates of that name.
+				if templateName != "" && templateName != template.Name {
+					logging.V(10).Infof("skipping template %q", template.Name)
+					continue
+				}
+
+				logging.V(10).Infof("adding template %q", template.Name)
+				s.addTemplate(Template{
+					name:               template.Name,
+					description:        "", // No description present.
```

Comment:
Will org templates never support descriptions?  Otherwise we should probably future proof this.